Switch helloworld-go with kn (#2807)
knative/docs: docs/serving/samples/hello-world/helloworld-go/index.md

type: "docs"
---

A simple web app written in Go that you can use for testing. It reads in an env
variable `TARGET` and prints `Hello ${TARGET}!`. If `TARGET` is not specified,
it will use `World` as the `TARGET`.

Follow the steps below to create the sample code and then deploy the app to your
cluster. You can also download a working copy of the sample, by running the
following commands:

```shell
git clone -b "{{< branch >}}" https://github.com/knative/docs knative-docs
cd knative-docs/docs/serving/samples/hello-world/helloworld-go
```

> ## Before you begin

Reviewer: +1 for the same document structure across all examples. The suggestion looks good to me.

> ## Before you begin

Author: I followed the headings from @rhuss PR, which I now realize are difference from here.  +1 for consistency, I just need to know what we want the shape to be 😅

   # Use the official Alpine image for a lean production container.
   # https://hub.docker.com/_/alpine
   # https://docs.docker.com/develop/develop-images/multistage-build/#use-multi-stage-builds
   FROM alpine:3
   RUN apk add --no-cache ca-certificates

   # Copy the binary to the production image from the builder stage.
   COPY --from=builder /app/server /server

   # Run the web service on container startup.
   CMD ["/server"]
   ```

1. Use the go tool to create a
   [`go.mod`](https://github.com/golang/go/wiki/Modules#gomod) manifest.

   ```shell
   go mod init github.com/knative/docs/docs/serving/samples/hello-world/helloworld-go
   ```

## Building and deploying the sample

Once you have recreated the sample code files (or used the files in the sample
folder) you're ready to build and deploy the sample app.

1. Use Docker to build the sample code into a container. To build and push with
   Docker Hub, run these commands replacing `{username}` with your Docker Hub
   username:

   ```shell
   # Build the container on your local machine
   docker build -t {username}/helloworld-go .

   # Push the container to docker registry

>    docker push {username}/helloworld-go

Author: Ack, we probably need to update the other samples.

   ```

1. After the build has completed and the container is pushed to docker hub, you
   can deploy the app into your cluster.  Pick one of the following methods:

   {{< tabs name="helloworld_go" default="kn" >}}
   {{% tab name="yaml" %}}

   Create a new file, `service.yaml` and copy the following service definition
   into the file. Make sure to replace `{username}` with your Docker Hub
   username.

   ```yaml
   apiVersion: serving.knative.dev/v1
   kind: Service
   metadata:
     name: helloworld-go
     namespace: default
   spec:
     template:
       spec:
         containers:
           - image: docker.io/{username}/helloworld-go
             env:
               - name: TARGET
                 value: "Go Sample v1"
   ```

>    Ensure that the container image value
>    in `service.yaml` matches the container you built in the previous step. Apply
>    the configuration using `kubectl`:
> 
>    ```shell
>    kubectl apply --filename service.yaml
>    ```

Reviewer 2: ```suggestion
   Check that the container image value in the `service.yaml` file matches the container you built in the previous step.
   
   Apply the configuration using `kubectl`:

   ```shell
   kubectl apply --filename service.yaml
   ```
```

Reviewer 2: No. 3 would be "apply..."

>    Now that your service is created, Knative will perform the following steps:

Reviewer 2: ```suggestion
   After your service is created, Knative will perform the following steps:
```

   - Create a new immutable revision for this version of the app.
   - Network programming to create a route, ingress, service, and load balance
     for your app.
   - Automatically scale your pods up and down (including to zero active pods).

   Run the following command to find the domain URL for your service:

   ```shell
   kubectl get ksvc helloworld-go  --output=custom-columns=NAME:.metadata.name,URL:.status.url
   ```

   Example:

   ```shell
    NAME                URL
    helloworld-go       http://helloworld-go.default.1.2.3.4.xip.io

>    ```

Reviewer 2: No. 4; also I feel like this should be a verification step instead.
I'm not sure that tabs are the best way to display this info at all tbh...

Author: What do you suggest?  We'd been talking about tabs for a while, which is why I'd bit the bullet here.  Definitely open to suggestions.

   {{< /tab >}}
   {{% tab name="kn" %}}

>    If using `kn` to deploy:

Reviewer 2: ```suggestion
   Use `kn` to deploy the service:
```

   ```shell
   kn service create helloworld-go --image=docker.io/{username}/helloworld-go --env TARGET="Go Sample v1"
   ```

>    This will block until your service is deployed and ready, and ultimately it will print the URL through which you can access the service.

Reviewer 2: would be much better to just show an expected example output, e.g, command for how they get the URL and a code block that shows a sample.

Author: Added sample output.

   {{< /tab >}}
   {{< /tabs >}}

> 1. Now you can make a request to your app and see the result. Replace

Reviewer 2: This should have a new heading and should be a section containing the verification steps, IMHO

Author: Added `## Verifying` above this.

type: "docs"
---

This guide describes the steps required to to create the `helloworld-go` sample app
and deploy it to your cluster.
The sample app reads a `TARGET` environment variable, and prints `Hello ${TARGET}!`.
If `TARGET` is not specified, `World` is used as the default value.

## Prerequisites

You will need:
- A Kubernetes cluster with [Knative installed and DNS configured](../../../../install/README.md).
- [Docker](https://www.docker.com) installed and running on your local machine, and a Docker Hub account configured.

> - (optional) The Knatice CLI client [kn](https://github.com/knative/client/releases) that simplifies the deployment. Alternative you can also use `kubectl` and apply resource files directly.

Reviewer 2: ```suggestion
- Optional: You can use the Knative CLI client [`kn`](https://github.com/knative/client/releases) to simplify resource creation and deployment. Alternatively, you can use `kubectl` to apply resource files directly.
```

Author: Fixed here and in the shell sample I copied it from.  thanks!

> ## Build

Reviewer 2: ```suggestion
## Building
```

We use deploying further down so I'd make them all gerunds or none of them

Author: Same, fixed both.
